### PR TITLE
Implementation of UDP hole punch transport

### LINKED ISSUE #394: Implement UDP holepunch transport
**Feature description**

We need to implement a UDP holepunch transport. This ticket depends on #391 
UDP holepunch transports should be more versatile than TCP holepunching.

### PR DESCRIPTION
**Pull request by nkryuchkov**
Did you run `make format && make check`? Yes

Fixes #394

 Changes:	
- Implementation of UDP hole punch transport
- Partial refactoring of direct transports
- For transports using `address-resolver`, if visors use the same IP, they try to establish a connection via local network first.

How to test this PR:
- Start generic integration env
- Run `make integration-startup-sudph`

### CODE REVIEW

File: pkg/snet/arclient/client.go [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // Package arclient implements address resolver client
+// TODO(nkryuchkov): move to internal
```
Comment (Darkren) [inline pkg/snet/arclient/client.go:2]: do you plan to move it in a separate PR?

File: pkg/snet/arclient/client.go [hunk 2]
```diff
@@ -325,11 +410,121 @@ func (c *httpClient) WS(ctx context.Context, dialCh <-chan cipher.PubKey) (<-cha
...
+
+	c.sudphAddrCh = addrCh
+
+	return nil
+}
+
+func (c *client) Close() error {
+	defer func() {
+		c.stcphConn = nil
+		// TODO(nkryuchkov): uncomment
```
Comment (Darkren) [inline pkg/snet/arclient/client.go:496]: Was it forgotten? As well as closing of sudph conn below

File: pkg/snet/directtransport/client.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"context"
+
+	"github.com/SkycoinProject/dmsg/cipher"
+	"github.com/SkycoinProject/skycoin/src/util/logging"
+)
+
+// Client is a direct transport client.
+type Client interface { // TODO(nkryuchkov): use
```
Comment (Darkren) [inline pkg/snet/directtransport/client.go]: Should these todos be removed?

File: pkg/snet/network.go [hunk 4]
```diff
@@ -605,6 +787,22 @@ func (n *Network) Listen(network string, port uint16) (*Listener, error) {
 			return nil, err
 		}
 
+		return makeListener(lis, network), nil
+	case sudpr.Type:
+		sudprC := n.clients.SudprC()
```
Comment (Darkren) [inline pkg/snet/network.go:727]: Can't client be `nil` here, like with stcph above?

File: pkg/snet/sudph/client.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+		c.mx.Lock()
+		defer c.mx.Unlock()
+
+		if err := c.addressResolver.Close(); err != nil {
+			c.log.WithError(err).Warnf("Failed to close address resolver client")
+		}
+
+		for _, lis := range c.lMap {
+			_ = lis.Close() // nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/sudph/client.go:409]: Why won't we log errors?

File: pkg/snet/sudpr/client.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+	log *logging.Logger
+
+	lPK             cipher.PubKey
+	lSK             cipher.SecKey
+	p               *porter.Porter
+	addressResolver arclient.APIClient
+	localAddr       string
+
+	lUDP net.Listener
+	lMap map[uint16]*directtransport.Listener // key: lPort
```
Comment (Darkren) [inline pkg/snet/sudpr/client.go:36]: Might be a matter of taste, but what do you think about calling this `listeners`? might be longer but not that much. it just feels to me like calling map a map is too explicit lol

**Review by Darkren — COMMENTED** [review]
Few minor questions, some may be of a taste. But anyway, a huge job is done, great!

File: pkg/snet/arclient/client.go (hunk 1 above)
Comment (nkryuchkov) [reply pkg/snet/arclient/client.go:2]: @Darkren Yes, I'm going to do that in a PR for https://github.com/SkycoinProject/skywire-mainnet/issues/409

File: pkg/snet/arclient/client.go (hunk 2 above)
Comment (nkryuchkov) [reply pkg/snet/arclient/client.go:496]: @Darkren No, I left it for https://github.com/SkycoinProject/skywire-mainnet/issues/409, because I'm going to combine all direct transports in one and I also need to add redialing logic.

File: pkg/snet/directtransport/client.go (hunk 3 above)
Comment (nkryuchkov) [reply pkg/snet/directtransport/client.go]: @Darkren I started to use this interface in a branch that fixes #409 and removed the `TODO` there.

File: pkg/snet/network.go (hunk 4 above)
Comment (nkryuchkov) [reply pkg/snet/network.go:727]: @Darkren Well spotted, thanks.

File: pkg/snet/sudph/client.go (hunk 5 above)
Comment (nkryuchkov) [reply pkg/snet/sudph/client.go:409]: I copied this code from `stcp` and did no changes to it because I was going to combine all transports in one. I agree that we should have an error check here, but would you mind if this would be done in a PR for #409? This code won't be duplicated 6 times there.

File: pkg/snet/sudpr/client.go (hunk 6 above)
Comment (nkryuchkov) [reply pkg/snet/sudpr/client.go:36]: @Darkren Good idea. To not rename this in 6 clients, do you mind if I do that in #409?

File: pkg/snet/sudph/client.go (hunk 5 above)
Comment (Darkren) [reply pkg/snet/sudph/client.go:409]: Of course, do as comfortable

File: pkg/snet/sudpr/client.go (hunk 6 above)
Comment (Darkren) [reply pkg/snet/sudpr/client.go:36]: Sure

**Review by Darkren — APPROVED** [review]
